### PR TITLE
Adding functionality to get LLH for various Asimov points.

### PR DESCRIPTION
**Pull request by kasettisiva**
Added/Implemented SetAsimov methods which will allow getting T2K LLH for various Asimov (true osc. parameters) points, not just Asimov 1, 2, 3, etc.

### CODE REVIEW

**Comment by cjbacchus** [comment]
We will need corresponding changes in `SendLLHOverBifrost()` so that the code inside the container knows what to do with these additional parameters.

**Comment by kasettisiva** [comment]
I modified the code a bit to include your comments.
Here is how it looks.

/// For use from inside the container
void SendLLHOverBifrost(Bifrost& bf, TemplateLLHGetter* llh)
{ 
  while(true){
    //send asimov osc pars
    OscPars asimovOscPars;
    int gotSentAsimovOsc;
    
    bf >> gotSentAsimovOsc;
    
    if(gotSentAsimovOsc) {
      bf >> asimovOscPars;
      llh->SetAsimovOscParameters(asimovOscPars);
    }
    int nasimovsystvecs;
    bf >> nasimovsystvecs;
    
    for(int i = 0; i < nasimovsystvecs; ++i){
      std::vector<double> asimovsysts;
      bf >> asimovsysts;
      CovTypes iCov = static_cast<CovTypes>(i);
      llh->SetAsimovSystParameters(iCov, asimovsysts);
    }

    OscPars p;
    bf >> p;
    llh->SetOscParameters(p);

    int nsystvecs;
    bf >> nsystvecs;

    for(int i = 0; i < nsystvecs; ++i){
      std::vector<double> systs;
      bf >> systs;
      CovTypes iCov = static_cast<CovTypes>(i);
      llh->SetParameters(iCov, systs);
    }

    bf << llh->GetLikelihood();
    }
}

File: LLHInterface.h [hunk 1]
```diff
@@ -60,8 +80,30 @@ class ReceiveLLHOverBifrost : public TemplateLLHGetter
...
+
   double GetLikelihood()
   {
+
+    fBF << int(fSendAsimovOsc);
+
+    if(fSendAsimovOsc) fBF << fAsimovOscPars;
+    int syst_size = fAsimovSysts.size();
+
+    fBF << syst_size;
```
Comment (cjbacchus) [inline LLHInterface.h]: I would combine these lines, just so we don't have this extra variable floating around.
```suggestion
    fBF << int(fAsimovSysts.size());
```

File: LLHInterface.h [hunk 2]
```diff
@@ -70,12 +112,19 @@ class ReceiveLLHOverBifrost : public TemplateLLHGetter
     for(const std::vector<double>& s: fSysts) fBF << s;
     double llh;
     fBF >> llh;
+
+    fSendAsimovOsc = false;
+    fAsimovSysts.clear();
```
Comment (cjbacchus) [inline LLHInterface.h]: I know I suggested this clearing behaviour, but it has the potential to confuse the user, if they expect to be able to set the systs, get a likelihood, and then make a small change to the systs. Maybe we should just have a `fSendAsimovSysts` too. I would still just send a length of zero for the case where it's false though.

File: LLHInterface.h [hunk 3]
```diff
@@ -21,6 +21,26 @@ Bifrost& operator<<(Bifrost& bf, OscPars& p)
...
+    //send asimov osc pars
+    OscPars asimovOscPars;
+    int gotSentAsimovOsc;
+
+    bf >> gotSentAsimovOsc;
+
+    if(gotSentAsimovOsc) {
+      bf >> asimovOscPars;
+      llh->SetAsimovOscParameters(asimovOscPars);
+    }
```
Comment (cjbacchus) [inline LLHInterface.h]: Move the OscPars to the scope they are actually needed
```suggestion
    int gotSentAsimovOsc;

    bf >> gotSentAsimovOsc;

    if(gotSentAsimovOsc) {
      OscPars asimovOscPars;
      bf >> asimovOscPars;
      llh->SetAsimovOscParameters(asimovOscPars);
    }
```

**Review by cjbacchus — COMMENTED** [review]
Have you been able to test this?

File: LLHInterface.h [hunk 4]
```diff
@@ -21,6 +21,29 @@ Bifrost& operator<<(Bifrost& bf, OscPars& p)
...
+    bf >> gotSentAsimovOsc;
+
+    if(gotSentAsimovOsc) {
+      OscPars asimovOscPars;
+      bf >> asimovOscPars;
+      llh->SetAsimovOscParameters(asimovOscPars);
+    }
+
+    int gotSentAsimovSysts;
+    bf >> gotSentAsimovSysts;
```
Comment (cjbacchus) [inline LLHInterface.h]: you aren't sending this anymore
```suggestion
```

File: LLHInterface.h [hunk 5]
```diff
@@ -70,12 +117,20 @@ class ReceiveLLHOverBifrost : public TemplateLLHGetter
     for(const std::vector<double>& s: fSysts) fBF << s;
     double llh;
     fBF >> llh;
+
+    fSendAsimovOsc = fSendAsimovSysts = false;
+    //fAsimovSysts.clear();
```
Comment (cjbacchus) [inline LLHInterface.h]: Don't keep old unneeded code
```suggestion
```

**Review by cjbacchus — COMMENTED** [review]
With these final comments this looks like it would potentially work.

Do you have a plan to test this? Reading the code and thinking only gets you so far.

Once we can confirm that it works, we need to think about migration plans without disrupting everyone. We will need asimov-capable containers (even if they just abort with an error when you try to set the asimov params) before users can update the version of bifrost outside, and we don't want to interrupt any of the current work.

**Comment by cjbacchus** [comment]
Tagged master as `v4` to reflect that it's the last version that matches the NOvA `v4_fixcosmicsrock` container.

**Pull request merged**